src/operator/c_lapack_api.h

```diff
@@ -160,7 +162,73 @@ inline void flip<cpu, double>(int m, int n,
 }
 
 
-#if MXNET_USE_LAPACK
+#if (MSHADOW_USE_MKL && MXNET_USE_LAPACK)
+
+  // We interface with the C-interface of MKL 
+  // as this is the preferred way. 
+  #include <mkl_lapacke.h>
+
+  #define MXNET_LAPACK_ROW_MAJOR LAPACK_ROW_MAJOR
+  #define MXNET_LAPACK_COL_MAJOR LAPACK_COL_MAJOR
+  #define MXNET_LAPACK_spotrf LAPACKE_spotrf 
+  #define MXNET_LAPACK_dpotrf LAPACKE_dpotrf 
+  #define MXNET_LAPACK_spotri LAPACKE_spotri 
+  #define MXNET_LAPACK_dpotri LAPACKE_dpotri 
+  #define mxnet_lapack_sposv  LAPACKE_sposv 
+  #define mxnet_lapack_dposv  LAPACKE_dposv 
+
+  // The following functions differ in signature from the 
+  // MXNET_LAPACK-signature and have to be wrapped.
+  #define MXNET_LAPACK_CWRAP_GELQF(prefix, dtype) \
+  inline int MXNET_LAPACK_##prefix##gelqf(int matrix_layout, int m, int n, \
+                                          dtype *a, int lda, dtype* tau, \
+                                          dtype* work, int lwork) { \
+    if (lwork != -1) { \
+      return LAPACKE_##prefix##gelqf(matrix_layout, m, n, a, lda, tau); \
+    } \
+    *work = 0; \
+    return 0; \
+  } 
+  MXNET_LAPACK_CWRAP_GELQF(s, float)
+  MXNET_LAPACK_CWRAP_GELQF(d, double)
+
+  #define MXNET_LAPACK_CWRAP_ORGLQ(prefix, dtype) \
+  inline int MXNET_LAPACK_##prefix##orglq(int matrix_layout, int m, int n, \
+                                          dtype *a, int lda, dtype* tau, \
+                                          dtype* work, int lwork) { \
+    if (lwork != -1) { \
+      return LAPACKE_##prefix##orglq(matrix_layout, m, n, m, a, lda, tau); \
+    } \
+    *work = 0; \
+    return 0; \
+  } 
+  MXNET_LAPACK_CWRAP_ORGLQ(s, float)
+  MXNET_LAPACK_CWRAP_ORGLQ(d, double)
+
+  // This has to be called internally in COL_MAJOR format even when matrix_layout
+  // is row-major as otherwise the eigenvectors would be returned as cols in a 
+  // row-major matrix layout (see MKL documentation). 
+  // We also have to allocate at least one DType element as workspace as the 
+  // calling code assumes that the workspace has at least that size. 
+  #define MXNET_LAPACK_CWRAP_SYEVD(prefix, dtype) \
+  inline int MXNET_LAPACK_##prefix##syevd(int matrix_layout, char uplo, int n, dtype *a, \
+                                          int lda, dtype *w, dtype *work, int lwork, \
+                                          int *iwork, int liwork) { \
+    if (lwork != -1) { \
+      char o(loup(uplo, (matrix_layout == MXNET_LAPACK_ROW_MAJOR))); \
+      return LAPACKE_##prefix##syevd(LAPACK_COL_MAJOR, 'V', o, n, a, lda, w); \
```

Comment:
Yep, I'd fix it to column major everywhere (not just here)!

Reply:
formally not needed. just here as MKL has this funny behaviour on output the the columns of the matrix are always the eigenvetors, regardless of matrix layout. 
Only needed if we think that the real intel MKL version is internally build doing unnecessary transpose-operations when matrix layout is row-major. There are lapacke-headers on github that seem to indicate that but  I don't think that Intel will put such inefficiencies in a final release of there claimed to be most efficient math-kernel-library.